src/app/server/Dnssd.cpp

```diff
@@ -343,6 +343,25 @@ void DnssdServer::StartServer()
     return StartServer(mode);
 }
 
+void DnssdServer::StopServer()
+{
+    DeviceLayer::PlatformMgr().RemoveEventHandler(OnPlatformEventWrapper, 0);
+
+    auto err = Dnssd::ServiceAdvertiser::Instance().RemoveServices();
```

Comment:
Is this safe?  This comes after `Dnssd::Resolver::Instance().Shutdown()` has been called...

Reply:
As in, should this StopServer() call be in `DeviceControllerSystemState::Shutdown` as opposed to after shutdown of the system state finishes?

Reply:
I have added a `virtual bool IsInitialized()` method to the `ServiceAdvertiser` to check if it is running before. Let me know if that is good enough.

Reply:
I'm not sure whether it is.  The problem is not that Dnssd::ServiceAdvertiser::Instance() might not be initialized.  The problem is that RemoveServices might depend on things that are shut down when Dnssd::Resolver::Instance() is shut down.  Can you check whether it is, please?

The "is initialized" check is not amazing, since init is async, so there may be things to remove even if we're not initialized yet, right?

Reply:
> I assume you are speaking about the platform implementation case where both the Resolver and the ServiceAdvertiser instance are a common instance ?

Yes.

> For those, the IsInitialized() flag is shared per my understanding, so if one is shutted down it should be reflected in the other as well ? 

I guess my point is this: If it _is_ shut down at this point, so that we did not call RemoveServices before it shut down, is that ok?  As in, is shutdown expected to do the equivalent of RemoveServices anyway?

Reply:
| I guess my point is this: If it is shut down at this point, so that we did not call RemoveServices before it shut down, is that ok? As in, is shutdown expected to do the equivalent of RemoveServices anyway?

I would say that shutting down dns-sd should do the equivalent of `RemoveServices` yes.